### DISCUSSION TITLE
Anti-Spam Delivery rules Definitions

### DISCUSSION THREAD

**Discussion by githubib** [mgeeky/decode-spam-headers#11] [Q&A]
Hello,
Just starting to use this tool.  I think it has a ton of potential.  What are the means by which we can decode these rule codes?  I see that you have decoded quite a few of them.  But I wonder what are the methods you employed?  Perhaps, we can duplicate these methods and decode a few of these as a community.  
Thanks!

**Comment by mgeeky** [comment]
Hello @githubib !

Thanks for the kind words. So, actually I followed a brutal, error-prone, vodoo-like methodology of sending tens of emails varying with tiny bits and then pulling SMTP headers on the recipient side. I have a luxury of having an email account in a really tighly locked environment.

To come up with that list I was sending emails towards my Office365 tighly locked account from various senders:
- Gsuite based sender
- Private postfix based sender
- phishing toolkit senders going directly and indirectly (through next-hop SMTP relays like GSuite)
- Hotmail 
- etc

To build up a corpora of emails I ended up sending variations of the below:
- empty emails, 
- emails containing phishy subjects, 
- emails with benign subject but phishy words in body
- emails with bening attachments
- emails with rogue attachments
- emails with bening and rogue URL 
- emails with URLs pointing at EXE/DLL/VBS/JS/LNK/HTML/PNG/ZIP/ISO/RAR (1 email = 1 url with 1 extension)

Having generated a corpora of N emails I then automated their delivery over a simple python script and pulled all of their headers on the recipient side (had to dig them out from Junk several times).

Then, with such a sample set of SMTP headers I used the script provided in this repo, which traversed directory with header text files, extracted "rule numbers" and then followed some naive corellation-like logic to filter out ones that weren't found in other emails.

Having identified a rule number not present in other emails, it was easier to isolate the difference of that email compared to others and then apply protein-based reasoning "why the fuck its in there?"

Hopefully you get an idea how much _sensible_ that can get :)

Let me know if all the above makes any sense!

EDIT: Actually, the more I think about it, the more I could see it nicely automated next time I'd sit down to deciphering their rules:

1. Auto-Generate corpus of 1000+ emails varying in little bits
2. Auto-Send them via a straightforward Python script
3. Connect to O365 account via IMAP-driven Python script, pull all the emails and save them as *.EMLs
4. Extract rule numbers
5. Apply black-box reasoning/correlation algorithm, or maybe just statistical package of some sort
6. Hunt for those occuring once or rarely
7. Identify why specific email was matched with particular rarely occuring rule.


Regards,
Mariusz.

**Comment by githubib** [comment]
Sounds like you stared at the Matrix green screen until you started seeing people moving!  Your software is a a little bit of an oasis in a cesspool of MSFT misinfo and non-support!  Thank you!

**Reply by mgeeky** [reply]
I stopped the first moment I saw a woman in red dress O _ O

My pleasure!
